content/codes/components/enigma-path-svg.ts

```diff
@@ -0,0 +1,612 @@
+		  (dasharray ? '" stroke-dasharray="' + dasharray.join(" ") : '') + 
+		  (dashoffset ? '" stroke-dashoffset="' + dashoffset : '') + 
+		  '" fill="' + fill + '"></path>';
+	}
+	
+	function pathpts(points:any[], close:boolean = false)
+	{
+	  return 'M' + points.map(p => p.x + ',' + p.y).join('L') + (close ? 'Z' : '');
+	}
+	
+	function addRoundRect(layer:SVGGElement, x:number, y:number, width:number, height:number, r_l:number, r_r:number, 
+		stroke:string, fill:string)
+	{
+		var path = '<path d="';
+		path += move(x + r_l, y);
+		path += line(x + width - r_r, y);
+		path += arc(r_r, x + width, y + r_r);
+		path += line(x + width, y + height - r_r);
+		path += arc(r_r, x + width - r_r, y + height);
+		path += line(x + r_l, y + height);
+		path += arc(r_l, x, y + height - r_l);
+		path += line(x, y + r_l);
+		path += arc(r_l, x + r_l, y);
+		path += '" stroke-width="1" stroke="' + stroke + '" fill="url(#' + fill + ')" />';
+		layer.innerHTML += path;
+	}
+
+	const svgdoc:Document = document.getElementById(svgid) as unknown as Document;
+
+	const incolour = "rgb(205, 14, 102)";
+	const outcolour = "rgb(34, 171, 36)";
+	const refcolour = "rgb(15, 130, 242)";
+	
+	const refl_top_col = "rgb(245, 226, 204)";
+	const refl_bot_col = "rgb(178, 160, 143)";
+	
+	const rotor_top_col = "rgb(210, 205, 199)";
+	const rotor_bot_col = "rgb(161, 157, 153)";
+	
+	const plug_top_col = "rgb(204, 226, 245)";
+	const plug_bot_col = "rgb(143, 160, 178)";
+	
+	const label_outer_col = "rgb(160, 160, 160)";
+	const label_inner_col = "rgb(240, 240, 240)";
+
+	const defs = svgdoc.getElementById(svgid + "_defs") as HTMLElement;
+	const styles = svgdoc.getElementById(svgid + "_styles") as HTMLElement;
+	const groups = svgdoc.getElementsByTagName("g");
+	const content = groups[0];
+	const encpath = groups[1];
+	const animpath = groups[2];
+
+	defs.innerHTML += makeLinearGradient("grad_refl", [0, 0, 0, 1], 
+		[{offset:0, col:refl_top_col},
+		{offset:1, col:refl_bot_col}]);
+
+	defs.innerHTML += makeLinearGradient("grad_rotor", [0, 0, 0, 1],
+		[{offset:0, col:rotor_top_col},
+		{offset:1, col:rotor_bot_col}]);
+
+	defs.innerHTML += makeLinearGradient("grad_rotor_label", [0, 0, 0, 1], 
+		[{offset:0, col:label_outer_col},
+		{offset:0.33, col:label_inner_col},
+		{offset:0.67, col:label_outer_col}]);
+
+	defs.innerHTML += makeLinearGradient("grad_plug", [0, 0, 0, 1], 
+		[{offset:0, col:plug_top_col},
+		{offset:1, col:plug_bot_col}]);
+			
+	// setup the text styles
+
+	styles.innerHTML += 
+		".heading { font: normal 14px sans-serif; fill:black;}" +
+		".normal { font: normal 12px sans-serif; fill:black; }" +
+		".inletter { font: normal 12px sans-serif; fill:" + incolour + "; }" +
+		".outletter { font: normal 12px sans-serif; fill:" + outcolour + "; }" +
+		".keyletter { font: normal 12px sans-serif; fill:#D0D0D0; }" +
+		".inkey { font: normal 12px sans-serif; fill:" + incolour + "; }" +
+		".outkey { font: normal 12px sans-serif; fill:" + outcolour + "; }";
```

Comment:
We should never set `.innerHTML` directly – especially if this code could just live in the corresponding `.less` file.